src/documentation/pipelines/test-your-pipeline.md

```diff
@@ -5,4 +5,263 @@ permalink: /documentation/pipelines/test-your-pipeline/
 ---
 # Test Your Pipeline
 
-> **Note:** There is an open JIRA issue to create this guide ([BEAM-901](https://issues.apache.org/jira/browse/BEAM-901)).
+* TOC
+{:toc}
+
+Testing your pipeline is a particularly important step in developing an effective data processing solution. The indirect nature of the Beam model, in which your user code constructs a pipeline graph to be executed remotely, can make debugging-failed runs a non-trivial task. Often it is faster and simpler to perform local unit testing on your pipeline code than to debug a pipeline's remote execution.
+
+Before running your pipeline on the service of your choice, unit testing your pipeline code locally is often the best way to identify and fix bugs in your pipeline code. Unit testing your pipeline locally also allows you to use your familiar/favorite local debugging tools.
+
+**Note:** When testing your code using the Google Cloud Dataflow service, consider limiting the number of worker instances for your pipeline to 1, or the minimum number appropriate for your test. Limiting the number of worker instances used during repeated test runs can provide significant time and cost savings. You can limit the number of workers your pipeline uses during test runs by setting the `--maxNumWorkers` execution option when you run your test pipeline.
```

Comment:
Replace with a comment about how after you develop against the DirectRunner, you'll like want to use your runner of choice to test on a small scale. For example the Flink runner can be used with a local Flink cluster or a remote Flink cluster.